src/test/coll.rs

```diff
@@ -754,3 +756,158 @@ async fn no_read_preference_to_standalone() {
 
     assert!(!command_started.command.contains_key("$readPreference"));
 }
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+struct UserType {
+    x: i32,
+    str: String,
+}
+
+#[cfg_attr(feature = "tokio-runtime", tokio::test)]
+#[cfg_attr(feature = "async-std-runtime", async_std::test)]
+#[function_name::named]
+async fn typed_insert_one() {
+    let _guard: RwLockReadGuard<()> = LOCK.run_concurrently().await;
+    let client = TestClient::new().await;
+
+    let coll = client
+        .init_db_and_typed_coll(function_name!(), function_name!())
+        .await;
+    let insert_data = UserType {
+        x: 1,
+        str: "a".into(),
+    };
+    insert_one_and_find(&coll, insert_data).await;
+
+    #[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+    struct OtherType {
+        b: bool,
+        nums: Vec<i32>,
+    }
+
+    let coll = coll.clone_with_type();
+    let insert_data = OtherType {
+        b: true,
+        nums: vec![1, 2, 3],
+    };
+    insert_one_and_find(&coll, insert_data).await;
+}
+
+async fn insert_one_and_find<T>(coll: &Collection<T>, insert_data: T)
+where
+    T: Serialize + DeserializeOwned + Clone + PartialEq + Debug,
+{
+    coll.insert_one(insert_data.clone(), None).await.unwrap();
+    let result = coll
+        .find_one(to_document(&insert_data).unwrap(), None)
+        .await
+        .unwrap();
+    match result {
+        Some(actual) => {
+            let actual: T = from_document(actual).unwrap();
```

Comment:
This part will presumably have to be updated in the next PR, right?

Reply:
yep, this step won't be necessary once returning `T` is implemented